session opening reminders timeout

Getting a few of these. This is more critical than closing reminders. Possibly due to large courses?
```
08:02:01.058
teammates.logic.automated.AutomatedRemindersServlet logMessage: TEAMMATESLOG|||feedbackSessionOpeningReminders|||send opening reminders|||true|||Auto|||Unknown|||Unknown|||Unknown|||Generating reminders for opening feedback sessions.|||/feedbackSessionOpeningReminders|||Unknown%2016082108020158
08:02:23.150
com.google.appengine.api.datastore.BaseQueryResultsSource logChunkSizeWarning: This query does not have a chunk size set in FetchOptions and has returned over 1000 results.  If result sets of this size are common for this query, consider setting a chunk size to improve performance.
  To disable this warning set the following system property in appengine-web.xml (the value of the property doesn't matter): 'appengine.datastore.disableChunkSizeWarning'
08:03:12.046
/feedbackSessionOpeningReminders
org.datanucleus.store.appengine.query.QueryExceptionWrappers$JDOQueryTimeoutException
	at org.datanucleus.store.appengine.query.QueryExceptionWrappers$2.get(QueryExceptionWrappers.java:88)
	at org.datanucleus.store.appengine.query.RuntimeExceptionWrappingIterator.hasNext(RuntimeExceptionWrappingIterator.java:105)
	at org.datanucleus.store.appengine.query.LazyResult$LazyAbstractListIterator.computeNext(LazyResult.java:210)
	at org.datanucleus.store.appengine.query.AbstractIterator.tryToComputeNext(AbstractIterator.java:132)
	at org.datanucleus.store.appengine.query.AbstractIterator.hasNext(AbstractIterator.java:127)
	at org.datanucleus.store.appengine.query.LazyResult$AbstractListIterator.hasNext(LazyResult.java:169)
	at teammates.storage.api.FeedbackSessionsDb.getFeedbackSessionsWithUnsentOpenEmail(FeedbackSessionsDb.java:182)
	at teammates.logic.core.FeedbackSessionsLogic.getFeedbackSessionsWhichNeedOpenEmailsToBeSent(FeedbackSessionsLogic.java:1117)
	at teammates.logic.core.FeedbackSessionsLogic.scheduleFeedbackSessionOpeningEmails(FeedbackSessionsLogic.java:1614)
	at teammates.logic.automated.FeedbackSessionOpeningRemindersServlet.doGet(FeedbackSessionOpeningRemindersServlet.java:21)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:617)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:717)
	at org.mortbay.jetty.servlet.ServletHolder.handle(ServletHolder.java:511)
	at org.mortbay.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1166)
	at teammates.storage.datastore.DatastoreFilter.doFilter(DatastoreFilter.java:29)
	at org.mortbay.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1157)
	at com.google.apphosting.utils.servlet.ParseBlobUploadFilter.doFilter(ParseBlobUploadFilter.java:125)
	at org.mortbay.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1157)
	at com.google.apphosting.runtime.jetty.SaveSessionFilter.doFilter(SaveSessionFilter.java:37)
	at org.mortbay.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1157)
	at com.google.apphosting.utils.servlet.JdbcMySqlConnectionCleanupFilter.doFilter(JdbcMySqlConnectionCleanupFilter.java:60)
	at org.mortbay.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1157)
	at com.google.apphosting.utils.servlet.TransactionCleanupFilter.doFilter(TransactionCleanupFilter.java:50)
	at org.mortbay.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1157)
	at org.mortbay.jetty.servlet.ServletHandler.handle(ServletHandler.java:388)
	at org.mortbay.jetty.security.SecurityHandler.handle(SecurityHandler.java:216)
	at org.mortbay.jetty.servlet.SessionHandler.handle(SessionHandler.java:182)
	at org.mortbay.jetty.handler.ContextHandler.handle(ContextHandler.java:765)
	at org.mortbay.jetty.webapp.WebAppContext.handle(WebAppContext.java:418)
	at com.google.apphosting.runtime.jetty.AppVersionHandlerMap.handle(AppVersionHandlerMap.java:260)
	at org.mortbay.jetty.handler.HandlerWrapper.handle(HandlerWrapper.java:152)
	at org.mortbay.jetty.Server.handle(Server.java:326)
	at org.mortbay.jetty.HttpConnection.handleRequest(HttpConnection.java:542)
	at org.mortbay.jetty.HttpConnection$RequestHandler.headerComplete(HttpConnection.java:923)
	at com.google.apphosting.runtime.jetty.RpcRequestParser.parseAvailable(RpcRequestParser.java:78)
	at org.mortbay.jetty.HttpConnection.handle(HttpConnection.java:404)
	at com.google.apphosting.runtime.jetty.JettyServletEngineAdapter.serviceRequest(JettyServletEngineAdapter.java:148)
	at com.google.apphosting.runtime.JavaRuntime$RequestRunnable.run(JavaRuntime.java:504)
	at com.google.tracing.TraceContext$TraceContextRunnable.runInContext(TraceContext.java:446)
	at com.google.tracing.TraceContext$TraceContextRunnable$1.run(TraceContext.java:453)
	at com.google.tracing.CurrentContext.runInContext(CurrentContext.java:276)
	at com.google.tracing.TraceContext$AbstractTraceContextCallback.runInInheritedContextNoUnref(TraceContext.java:312)
	at com.google.tracing.TraceContext$AbstractTraceContextCallback.runInInheritedContext(TraceContext.java:304)
	at com.google.tracing.TraceContext$TraceContextRunnable.run(TraceContext.java:450)
	at com.google.apphosting.runtime.ThreadGroupPool$PoolEntry.run(ThreadGroupPool.java:235)
	at java.lang.Thread.run(Thread.java:745)
Caused by: org.datanucleus.store.appengine.DatastoreExceptionTranslator$1: The datastore operation timed out, or the data was temporarily unavailable.
	at org.datanucleus.store.appengine.DatastoreExceptionTranslator.wrapDatastoreTimeoutExceptionForQuery(DatastoreExceptionTranslator.java:65)
	at org.datanucleus.store.appengine.query.QueryExceptionWrappers$1.get(QueryExceptionWrappers.java:55)
	at org.datanucleus.store.appengine.query.QueryExceptionWrappers$2.get(QueryExceptionWrappers.java:86)
	... 45 more
Caused by: com.google.appengine.api.datastore.DatastoreTimeoutException: The datastore operation timed out, or the data was temporarily unavailable.
	at com.google.appengine.api.datastore.DatastoreApiHelper.translateError(DatastoreApiHelper.java:64)
	at com.google.appengine.api.datastore.DatastoreApiHelper$1.convertException(DatastoreApiHelper.java:129)
	at com.google.appengine.api.utils.FutureWrapper.get(FutureWrapper.java:97)
	at com.google.appengine.api.datastore.FutureHelper.getInternal(FutureHelper.java:76)
	at com.google.appengine.api.datastore.FutureHelper.quietGet(FutureHelper.java:36)
	at com.google.appengine.api.datastore.BaseQueryResultsSource.loadMoreEntities(BaseQueryResultsSource.java:197)
	at com.google.appengine.api.datastore.BaseQueryResultsSource.loadMoreEntities(BaseQueryResultsSource.java:171)
	at com.google.appengine.api.datastore.QueryResultIteratorImpl.ensureLoaded(QueryResultIteratorImpl.java:148)
	at com.google.appengine.api.datastore.QueryResultIteratorImpl.hasNext(QueryResultIteratorImpl.java:64)
	at com.google.appengine.api.datastore.PreparedMultiQuery$FilteredMultiQueryIterator.computeNext(PreparedMultiQuery.java:249)
	at com.google.appengine.api.datastore.PreparedMultiQuery$FilteredMultiQueryIterator.computeNext(PreparedMultiQuery.java:185)
	at com.google.appengine.api.datastore.AbstractIterator.tryToComputeNext(AbstractIterator.java:46)
	at com.google.appengine.api.datastore.AbstractIterator.hasNext(AbstractIterator.java:41)
	at com.google.appengine.api.datastore.QueryResultIteratorDelegator.hasNext(QueryResultIteratorDelegator.java:38)
	at org.datanucleus.store.appengine.query.RuntimeExceptionWrappingIterator$1.get(RuntimeExceptionWrappingIterator.java:50)
	at org.datanucleus.store.appengine.query.RuntimeExceptionWrappingIterator$1.get(RuntimeExceptionWrappingIterator.java:46)
	at org.datanucleus.store.appengine.query.QueryExceptionWrappers$1.get(QueryExceptionWrappers.java:51)
	... 46 more

--- a/src/main/java/teammates/logic/core/FeedbackSessionsLogic.java
+++ b/src/main/java/teammates/logic/core/FeedbackSessionsLogic.java
@@ -1098,7 +1098,7 @@ public class FeedbackSessionsLogic {
      */
     public List<FeedbackSessionAttributes> getFeedbackSessionsWhichNeedAutomatedPublishedEmailsToBeSent() {
         List<FeedbackSessionAttributes> sessions =
-                fsDb.getFeedbackSessionsNeedingPublishedEmail();
+                fsDb.getFeedbackSessionsPossiblyNeedingPublishedEmail();
         List<FeedbackSessionAttributes> sessionsToSendEmailsFor =
                 new ArrayList<FeedbackSessionAttributes>();
 
@@ -1113,7 +1113,7 @@ public class FeedbackSessionsLogic {
 
     public List<FeedbackSessionAttributes> getFeedbackSessionsWhichNeedOpenEmailsToBeSent() {
         List<FeedbackSessionAttributes> sessions =
-                fsDb.getFeedbackSessionsNeedingOpenEmail();
+                fsDb.getFeedbackSessionsPossiblyNeedingOpenEmail();
         List<FeedbackSessionAttributes> sessionsToSendEmailsFor =
                 new ArrayList<FeedbackSessionAttributes>();
 
@@ -1622,7 +1622,7 @@ public class FeedbackSessionsLogic {
                 ArrayList<FeedbackSessionAttributes>();
 
         List<FeedbackSessionAttributes> nonPrivateSessions =
-                fsDb.getFeedbackSessionsNeedingClosingEmail();
+                fsDb.getFeedbackSessionsPossiblyNeedingClosingEmail();
 
         for (FeedbackSessionAttributes session : nonPrivateSessions) {
             if (session.getFeedbackSessionType() != FeedbackSessionType.PRIVATE
@@ -1640,7 +1640,7 @@ public class FeedbackSessionsLogic {
     public List<FeedbackSessionAttributes> getFeedbackSessionsClosedWithinThePastHour() {
         List<FeedbackSessionAttributes> requiredSessions = new ArrayList<FeedbackSessionAttributes>();
         List<FeedbackSessionAttributes> nonPrivateSessions =
-                fsDb.getFeedbackSessionsNeedingClosedEmail();
+                fsDb.getFeedbackSessionsPossiblyNeedingClosedEmail();
 
         for (FeedbackSessionAttributes session : nonPrivateSessions) {
             // is session closed in the past 1 hour

--- a/src/main/java/teammates/logic/core/FeedbackSessionsLogic.java
+++ b/src/main/java/teammates/logic/core/FeedbackSessionsLogic.java
@@ -1098,14 +1098,13 @@ public class FeedbackSessionsLogic {
      */
     public List<FeedbackSessionAttributes> getFeedbackSessionsWhichNeedAutomatedPublishedEmailsToBeSent() {
         List<FeedbackSessionAttributes> sessions =
-                fsDb.getFeedbackSessionsWithUnsentPublishedEmail();
+                fsDb.getFeedbackSessionsNeedingPublishedEmail();
         List<FeedbackSessionAttributes> sessionsToSendEmailsFor =
                 new ArrayList<FeedbackSessionAttributes>();
 
         for (FeedbackSessionAttributes session : sessions) {
             // automated emails are required only for custom publish times
-            if (session.isPublished() && session.isPublishedEmailEnabled()
-                    && !TimeHelper.isSpecialTime(session.getResultsVisibleFromTime())) {
+            if (session.isPublished() && !TimeHelper.isSpecialTime(session.getResultsVisibleFromTime())) {
                 sessionsToSendEmailsFor.add(session);
             }
         }
@@ -1114,12 +1113,12 @@ public class FeedbackSessionsLogic {
 
     public List<FeedbackSessionAttributes> getFeedbackSessionsWhichNeedOpenEmailsToBeSent() {
         List<FeedbackSessionAttributes> sessions =
-                fsDb.getFeedbackSessionsWithUnsentOpenEmail();
+                fsDb.getFeedbackSessionsNeedingOpenEmail();
         List<FeedbackSessionAttributes> sessionsToSendEmailsFor =
                 new ArrayList<FeedbackSessionAttributes>();
 
         for (FeedbackSessionAttributes session : sessions) {
-            if (session.isOpened()) {
+            if (session.getFeedbackSessionType() != FeedbackSessionType.PRIVATE && session.isOpened()) {
                 sessionsToSendEmailsFor.add(session);
             }
         }
@@ -1626,7 +1625,8 @@ public class FeedbackSessionsLogic {
                 fsDb.getFeedbackSessionsNeedingClosingEmail();
 
         for (FeedbackSessionAttributes session : nonPrivateSessions) {
-            if (session.isClosingWithinTimeLimit(SystemParams.NUMBER_OF_HOURS_BEFORE_CLOSING_ALERT)) {
+            if (session.getFeedbackSessionType() != FeedbackSessionType.PRIVATE
+                    && session.isClosingWithinTimeLimit(SystemParams.NUMBER_OF_HOURS_BEFORE_CLOSING_ALERT)) {
                 requiredSessions.add(session);
             }
         }
@@ -1644,7 +1644,8 @@ public class FeedbackSessionsLogic {
 
         for (FeedbackSessionAttributes session : nonPrivateSessions) {
             // is session closed in the past 1 hour
-            if (session.isClosedWithinPastHour()) {
+            if (session.getFeedbackSessionType() != FeedbackSessionType.PRIVATE
+                    && session.isClosedWithinPastHour()) {
                 requiredSessions.add(session);
             }
         }

--- a/src/main/java/teammates/storage/api/FeedbackSessionsDb.java
+++ b/src/main/java/teammates/storage/api/FeedbackSessionsDb.java
@@ -170,13 +170,11 @@ public class FeedbackSessionsDb extends EntitiesDb {
     }
         
     /**
-     * Preconditions: <br>
-     * * All parameters are non-null.
      * @return An empty list if no sessions are found that have unsent open emails.
      */
-    public List<FeedbackSessionAttributes> getFeedbackSessionsNeedingOpenEmail() {
+    public List<FeedbackSessionAttributes> getFeedbackSessionsPossiblyNeedingOpenEmail() {
                 
-        List<FeedbackSession> fsList = getFeedbackSessionEntitiesNeedingOpenEmail();
+        List<FeedbackSession> fsList = getFeedbackSessionEntitiesPossiblyNeedingOpenEmail();
         List<FeedbackSessionAttributes> fsaList = new ArrayList<FeedbackSessionAttributes>();
         
         for (FeedbackSession fs : fsList) {
@@ -190,9 +188,9 @@ public class FeedbackSessionsDb extends EntitiesDb {
     /**
      * @return An empty list if no sessions are found that have unsent closing emails.
      */
-    public List<FeedbackSessionAttributes> getFeedbackSessionsNeedingClosingEmail() {
+    public List<FeedbackSessionAttributes> getFeedbackSessionsPossiblyNeedingClosingEmail() {
                 
-        List<FeedbackSession> fsList = getFeedbackSessionEntitiesNeedingClosingEmail();
+        List<FeedbackSession> fsList = getFeedbackSessionEntitiesPossiblyNeedingClosingEmail();
         List<FeedbackSessionAttributes> fsaList = new ArrayList<FeedbackSessionAttributes>();
         
         for (FeedbackSession fs : fsList) {
@@ -206,9 +204,9 @@ public class FeedbackSessionsDb extends EntitiesDb {
     /**
      * @return An empty list if no sessions are found that have unsent closed emails.
      */
-    public List<FeedbackSessionAttributes> getFeedbackSessionsNeedingClosedEmail() {
+    public List<FeedbackSessionAttributes> getFeedbackSessionsPossiblyNeedingClosedEmail() {
                 
-        List<FeedbackSession> fsList = getFeedbackSessionEntitiesNeedingClosedEmail();
+        List<FeedbackSession> fsList = getFeedbackSessionEntitiesPossiblyNeedingClosedEmail();
         List<FeedbackSessionAttributes> fsaList = new ArrayList<FeedbackSessionAttributes>();
         
         for (FeedbackSession fs : fsList) {
@@ -220,13 +218,11 @@ public class FeedbackSessionsDb extends EntitiesDb {
     }
     
     /**
-     * Preconditions: <br>
-     * * All parameters are non-null.
      * @return An empty list if no sessions are found that have unsent published emails.
      */
-    public List<FeedbackSessionAttributes> getFeedbackSessionsNeedingPublishedEmail() {
+    public List<FeedbackSessionAttributes> getFeedbackSessionsPossiblyNeedingPublishedEmail() {
         
-        List<FeedbackSession> fsList = getFeedbackSessionEntitiesNeedingPublishedEmail();
+        List<FeedbackSession> fsList = getFeedbackSessionEntitiesPossiblyNeedingPublishedEmail();
         List<FeedbackSessionAttributes> fsaList = new ArrayList<FeedbackSessionAttributes>();
         
         for (FeedbackSession fs : fsList) {
@@ -541,40 +537,43 @@ public class FeedbackSessionsDb extends EntitiesDb {
     }
     
     @SuppressWarnings("unchecked")
-    private List<FeedbackSession> getFeedbackSessionEntitiesNeedingOpenEmail() {
+    private List<FeedbackSession> getFeedbackSessionEntitiesPossiblyNeedingOpenEmail() {
         Query q = getPm().newQuery(FeedbackSession.class);
         q.declareParameters("java.util.Date startTimeParam, boolean sentParam");
         q.setFilter("startTime > startTimeParam && sentOpenEmail == sentParam");
         
+        // only get sessions with startTime within the past two days to reduce the number of sessions returned
         Date d = TimeHelper.getDateOffsetToCurrentTime(-2);
         
         return (List<FeedbackSession>) q.execute(d, false);
     }
     
     @SuppressWarnings("unchecked")
-    private List<FeedbackSession> getFeedbackSessionEntitiesNeedingClosingEmail() {
+    private List<FeedbackSession> getFeedbackSessionEntitiesPossiblyNeedingClosingEmail() {
         Query q = getPm().newQuery(FeedbackSession.class);
         q.declareParameters("java.util.Date endTimeParam, boolean sentParam, boolean enableParam");
         q.setFilter("endTime > endTimeParam && sentClosingEmail == sentParam && isClosingEmailEnabled == enableParam");
         
+        // only get sessions with endTime within the past two days to reduce the number of sessions returned
         Date d = TimeHelper.getDateOffsetToCurrentTime(-2);
         
         return (List<FeedbackSession>) q.execute(d, false, true);
     }
     
     @SuppressWarnings("unchecked")
-    private List<FeedbackSession> getFeedbackSessionEntitiesNeedingClosedEmail() {
+    private List<FeedbackSession> getFeedbackSessionEntitiesPossiblyNeedingClosedEmail() {
         Query q = getPm().newQuery(FeedbackSession.class);
         q.declareParameters("java.util.Date endTimeParam, boolean sentParam, boolean enableParam");
         q.setFilter("endTime > endTimeParam && sentClosedEmail == sentParam && isClosingEmailEnabled == enableParam");
         
+        // only get sessions with endTime within the past two days to reduce the number of sessions returned
         Date d = TimeHelper.getDateOffsetToCurrentTime(-2);
         
         return (List<FeedbackSession>) q.execute(d, false, true);
     }
     
     @SuppressWarnings("unchecked")
-    private List<FeedbackSession> getFeedbackSessionEntitiesNeedingPublishedEmail() {
+    private List<FeedbackSession> getFeedbackSessionEntitiesPossiblyNeedingPublishedEmail() {
         Query q = getPm().newQuery(FeedbackSession.class);
         q.declareParameters("boolean sentParam, boolean enableParam, Enum notTypeParam");
         q.setFilter("sentPublishedEmail == sentParam && isPublishedEmailEnabled == enableParam "

--- a/src/main/java/teammates/storage/api/FeedbackSessionsDb.java
+++ b/src/main/java/teammates/storage/api/FeedbackSessionsDb.java
@@ -554,8 +554,8 @@ public class FeedbackSessionsDb extends EntitiesDb {
     @SuppressWarnings("unchecked")
     private List<FeedbackSession> getFeedbackSessionEntitiesNeedingClosingEmail() {
         Query q = getPm().newQuery(FeedbackSession.class);
-        q.declareParameters("java.util.Date startTimeParam, boolean sentParam, boolean enableParam");
-        q.setFilter("endTime > startTimeParam && sentClosingEmail == sentParam && isClosingEmailEnabled == enableParam");
+        q.declareParameters("java.util.Date endTimeParam, boolean sentParam, boolean enableParam");
+        q.setFilter("endTime > endTimeParam && sentClosingEmail == sentParam && isClosingEmailEnabled == enableParam");
         
         Date d = TimeHelper.getDateOffsetToCurrentTime(-2);
         
@@ -565,8 +565,8 @@ public class FeedbackSessionsDb extends EntitiesDb {
     @SuppressWarnings("unchecked")
     private List<FeedbackSession> getFeedbackSessionEntitiesNeedingClosedEmail() {
         Query q = getPm().newQuery(FeedbackSession.class);
-        q.declareParameters("java.util.Date startTimeParam, boolean sentParam, boolean enableParam");
-        q.setFilter("endTime > startTimeParam && sentClosedEmail == sentParam && isClosingEmailEnabled == enableParam");
+        q.declareParameters("java.util.Date endTimeParam, boolean sentParam, boolean enableParam");
+        q.setFilter("endTime > endTimeParam && sentClosedEmail == sentParam && isClosingEmailEnabled == enableParam");
         
         Date d = TimeHelper.getDateOffsetToCurrentTime(-2);
         

--- a/src/main/java/teammates/storage/api/FeedbackSessionsDb.java
+++ b/src/main/java/teammates/storage/api/FeedbackSessionsDb.java
@@ -543,8 +543,8 @@ public class FeedbackSessionsDb extends EntitiesDb {
     @SuppressWarnings("unchecked")
     private List<FeedbackSession> getFeedbackSessionEntitiesNeedingOpenEmail() {
         Query q = getPm().newQuery(FeedbackSession.class);
-        q.declareParameters("java.util.Date yesterday, boolean sentParam");
-        q.setFilter("startTime > yesterday && sentOpenEmail == sentParam");
+        q.declareParameters("java.util.Date startTimeParam, boolean sentParam");
+        q.setFilter("startTime > startTimeParam && sentOpenEmail == sentParam");
         
         Date d = TimeHelper.getDateOffsetToCurrentTime(-2);
         
@@ -554,8 +554,8 @@ public class FeedbackSessionsDb extends EntitiesDb {
     @SuppressWarnings("unchecked")
     private List<FeedbackSession> getFeedbackSessionEntitiesNeedingClosingEmail() {
         Query q = getPm().newQuery(FeedbackSession.class);
-        q.declareParameters("java.util.Date yesterday, boolean sentParam, boolean enableParam");
-        q.setFilter("endTime > yesterday && sentClosingEmail == sentParam && isClosingEmailEnabled == enableParam");
+        q.declareParameters("java.util.Date startTimeParam, boolean sentParam, boolean enableParam");
+        q.setFilter("endTime > startTimeParam && sentClosingEmail == sentParam && isClosingEmailEnabled == enableParam");
         
         Date d = TimeHelper.getDateOffsetToCurrentTime(-2);
         
@@ -565,8 +565,8 @@ public class FeedbackSessionsDb extends EntitiesDb {
     @SuppressWarnings("unchecked")
     private List<FeedbackSession> getFeedbackSessionEntitiesNeedingClosedEmail() {
         Query q = getPm().newQuery(FeedbackSession.class);
-        q.declareParameters("java.util.Date yesterday, boolean sentParam, boolean enableParam");
-        q.setFilter("endTime > yesterday && sentClosedEmail == sentParam && isClosingEmailEnabled == enableParam");
+        q.declareParameters("java.util.Date startTimeParam, boolean sentParam, boolean enableParam");
+        q.setFilter("endTime > startTimeParam && sentClosedEmail == sentParam && isClosingEmailEnabled == enableParam");
         
         Date d = TimeHelper.getDateOffsetToCurrentTime(-2);
         

--- a/src/main/java/teammates/storage/api/FeedbackSessionsDb.java
+++ b/src/main/java/teammates/storage/api/FeedbackSessionsDb.java
@@ -174,9 +174,9 @@ public class FeedbackSessionsDb extends EntitiesDb {
      * * All parameters are non-null.
      * @return An empty list if no sessions are found that have unsent open emails.
      */
-    public List<FeedbackSessionAttributes> getFeedbackSessionsWithUnsentOpenEmail() {
+    public List<FeedbackSessionAttributes> getFeedbackSessionsNeedingOpenEmail() {
                 
-        List<FeedbackSession> fsList = getFeedbackSessionEntitiesWithUnsentOpenEmail();
+        List<FeedbackSession> fsList = getFeedbackSessionEntitiesNeedingOpenEmail();
         List<FeedbackSessionAttributes> fsaList = new ArrayList<FeedbackSessionAttributes>();
         
         for (FeedbackSession fs : fsList) {
@@ -224,10 +224,9 @@ public class FeedbackSessionsDb extends EntitiesDb {
      * * All parameters are non-null.
      * @return An empty list if no sessions are found that have unsent published emails.
      */
-    public List<FeedbackSessionAttributes> getFeedbackSessionsWithUnsentPublishedEmail() {
+    public List<FeedbackSessionAttributes> getFeedbackSessionsNeedingPublishedEmail() {
         
-        
-        List<FeedbackSession> fsList = getFeedbackSessionEntitiesWithUnsentPublishedEmail();
+        List<FeedbackSession> fsList = getFeedbackSessionEntitiesNeedingPublishedEmail();
         List<FeedbackSessionAttributes> fsaList = new ArrayList<FeedbackSessionAttributes>();
         
         for (FeedbackSession fs : fsList) {
@@ -542,41 +541,46 @@ public class FeedbackSessionsDb extends EntitiesDb {
     }
     
     @SuppressWarnings("unchecked")
-    private List<FeedbackSession> getFeedbackSessionEntitiesWithUnsentOpenEmail() {
+    private List<FeedbackSession> getFeedbackSessionEntitiesNeedingOpenEmail() {
         Query q = getPm().newQuery(FeedbackSession.class);
-        q.declareParameters("boolean sentParam, Enum notTypeParam");
-        q.setFilter("sentOpenEmail == sentParam && feedbackSessionType != notTypeParam");
+        q.declareParameters("java.util.Date yesterday, boolean sentParam");
+        q.setFilter("startTime > yesterday && sentOpenEmail == sentParam");
         
-        return (List<FeedbackSession>) q.execute(false, FeedbackSessionType.PRIVATE);
+        Date d = TimeHelper.getDateOffsetToCurrentTime(-2);
+        
+        return (List<FeedbackSession>) q.execute(d, false);
     }
     
     @SuppressWarnings("unchecked")
     private List<FeedbackSession> getFeedbackSessionEntitiesNeedingClosingEmail() {
         Query q = getPm().newQuery(FeedbackSession.class);
-        q.declareParameters("boolean sentParam, boolean enableParam, Enum notTypeParam");
-        q.setFilter("sentClosingEmail == sentParam && isClosingEmailEnabled == enableParam "
-                    + "&& feedbackSessionType != notTypeParam");
+        q.declareParameters("java.util.Date yesterday, boolean sentParam, boolean enableParam");
+        q.setFilter("endTime > yesterday && sentClosingEmail == sentParam && isClosingEmailEnabled == enableParam");
         
-        return (List<FeedbackSession>) q.execute(false, true, FeedbackSessionType.PRIVATE);
+        Date d = TimeHelper.getDateOffsetToCurrentTime(-2);
+        
+        return (List<FeedbackSession>) q.execute(d, false, true);
     }
     
     @SuppressWarnings("unchecked")
     private List<FeedbackSession> getFeedbackSessionEntitiesNeedingClosedEmail() {
         Query q = getPm().newQuery(FeedbackSession.class);
-        q.declareParameters("boolean sentParam, boolean enableParam, Enum notTypeParam");
-        q.setFilter("sentClosedEmail == sentParam && isClosingEmailEnabled == enableParam "
-                    + "&& feedbackSessionType != notTypeParam");
+        q.declareParameters("java.util.Date yesterday, boolean sentParam, boolean enableParam");
+        q.setFilter("endTime > yesterday && sentClosedEmail == sentParam && isClosingEmailEnabled == enableParam");
         
-        return (List<FeedbackSession>) q.execute(false, true, FeedbackSessionType.PRIVATE);
+        Date d = TimeHelper.getDateOffsetToCurrentTime(-2);
+        
+        return (List<FeedbackSession>) q.execute(d, false, true);
     }
     
     @SuppressWarnings("unchecked")
-    private List<FeedbackSession> getFeedbackSessionEntitiesWithUnsentPublishedEmail() {
+    private List<FeedbackSession> getFeedbackSessionEntitiesNeedingPublishedEmail() {
         Query q = getPm().newQuery(FeedbackSession.class);
-        q.declareParameters("boolean sentParam, Enum notTypeParam");
-        q.setFilter("sentPublishedEmail == sentParam && feedbackSessionType != notTypeParam");
+        q.declareParameters("boolean sentParam, boolean enableParam, Enum notTypeParam");
+        q.setFilter("sentPublishedEmail == sentParam && isPublishedEmailEnabled == enableParam "
+                    + "&& feedbackSessionType != notTypeParam");
         
-        return (List<FeedbackSession>) q.execute(false, FeedbackSessionType.PRIVATE);
+        return (List<FeedbackSession>) q.execute(false, true, FeedbackSessionType.PRIVATE);
     }
     
     private FeedbackSession getFeedbackSessionEntity(String feedbackSessionName, String courseId) {

--- a/src/test/java/teammates/test/cases/automated/FeedbackSessionOpeningReminderTest.java
+++ b/src/test/java/teammates/test/cases/automated/FeedbackSessionOpeningReminderTest.java
@@ -84,7 +84,7 @@ public class FeedbackSessionOpeningReminderTest extends BaseComponentUsingTaskQu
         session1.setStartTime(TimeHelper.getDateOffsetToCurrentTime(2));
         session1.setEndTime(TimeHelper.getDateOffsetToCurrentTime(3));
         fsLogic.updateFeedbackSession(session1);
-        session1.setStartTime(TimeHelper.getDateOffsetToCurrentTime(-2));
+        session1.setStartTime(TimeHelper.getHoursOffsetToCurrentTime(-47));
         fsLogic.updateFeedbackSession(session1);
         verifyPresentInDatastore(session1);
         
@@ -98,7 +98,7 @@ public class FeedbackSessionOpeningReminderTest extends BaseComponentUsingTaskQu
         session2.setEndTime(TimeHelper.getDateOffsetToCurrentTime(3));
         session2.setOpeningEmailEnabled(false);
         fsLogic.updateFeedbackSession(session2);
-        session2.setStartTime(TimeHelper.getDateOffsetToCurrentTime(-2));
+        session2.setStartTime(TimeHelper.getHoursOffsetToCurrentTime(-47));
         fsLogic.updateFeedbackSession(session2);
         verifyPresentInDatastore(session2);
         
@@ -140,7 +140,7 @@ public class FeedbackSessionOpeningReminderTest extends BaseComponentUsingTaskQu
         session2.setEndTime(TimeHelper.getDateOffsetToCurrentTime(3));
         session2.setOpeningEmailEnabled(false);
         fsLogic.updateFeedbackSession(session2);
-        session2.setStartTime(TimeHelper.getDateOffsetToCurrentTime(-2));
+        session2.setStartTime(TimeHelper.getHoursOffsetToCurrentTime(-47));
         fsLogic.updateFeedbackSession(session2);
         
         int course1StudentCount = 5;

--- a/src/test/java/teammates/test/cases/logic/FeedbackSessionsLogicTest.java
+++ b/src/test/java/teammates/test/cases/logic/FeedbackSessionsLogicTest.java
@@ -192,7 +192,7 @@ public class FeedbackSessionsLogicTest extends BaseComponentTestCase {
         session.setTimeZone(0);
         session.setFeedbackSessionType(FeedbackSessionType.STANDARD);
         session.setSessionVisibleFromTime(TimeHelper.getDateOffsetToCurrentTime(-2));
-        session.setStartTime(TimeHelper.getDateOffsetToCurrentTime(-2));
+        session.setStartTime(TimeHelper.getHoursOffsetToCurrentTime(-47));
         session.setEndTime(TimeHelper.getDateOffsetToCurrentTime(1));
         session.setSentOpenEmail(false);
         fsLogic.createFeedbackSession(session);

--- a/src/test/java/teammates/test/cases/storage/FeedbackSessionsDbTest.java
+++ b/src/test/java/teammates/test/cases/storage/FeedbackSessionsDbTest.java
@@ -100,10 +100,10 @@ public class FeedbackSessionsDbTest extends BaseComponentTestCase {
 
         testGetFeedbackSessions();
         testGetFeedbackSessionsForCourse();
-        testGetFeedbackSessionsNeedingOpenEmail();
-        testGetFeedbackSessionsNeedingClosingEmail();
-        testGetFeedbackSessionsNeedingClosedEmail();
-        testGetFeedbackSessionsNeedingPublishedEmail();
+        testGetFeedbackSessionsPossiblyNeedingOpenEmail();
+        testGetFeedbackSessionsPossiblyNeedingClosingEmail();
+        testGetFeedbackSessionsPossiblyNeedingClosedEmail();
+        testGetFeedbackSessionsPossiblyNeedingPublishedEmail();
     }
     
     private void testGetFeedbackSessions() {
@@ -178,11 +178,11 @@ public class FeedbackSessionsDbTest extends BaseComponentTestCase {
         assertTrue(fsDb.getFeedbackSessionsForCourse("idOfCourseNoEvals").isEmpty());
     }
     
-    private void testGetFeedbackSessionsNeedingOpenEmail() {
+    private void testGetFeedbackSessionsPossiblyNeedingOpenEmail() {
         
         ______TS("standard success case");
         
-        List<FeedbackSessionAttributes> fsaList = fsDb.getFeedbackSessionsNeedingOpenEmail();
+        List<FeedbackSessionAttributes> fsaList = fsDb.getFeedbackSessionsPossiblyNeedingOpenEmail();
         
         assertEquals(1, fsaList.size());
         for (FeedbackSessionAttributes fsa : fsaList) {
@@ -191,11 +191,11 @@ public class FeedbackSessionsDbTest extends BaseComponentTestCase {
         
     }
     
-    private void testGetFeedbackSessionsNeedingClosingEmail() {
+    private void testGetFeedbackSessionsPossiblyNeedingClosingEmail() {
         
         ______TS("standard success case");
         
-        List<FeedbackSessionAttributes> fsaList = fsDb.getFeedbackSessionsNeedingClosingEmail();
+        List<FeedbackSessionAttributes> fsaList = fsDb.getFeedbackSessionsPossiblyNeedingClosingEmail();
         
         assertEquals(6, fsaList.size());
         for (FeedbackSessionAttributes fsa : fsaList) {
@@ -205,11 +205,11 @@ public class FeedbackSessionsDbTest extends BaseComponentTestCase {
         
     }
     
-    private void testGetFeedbackSessionsNeedingClosedEmail() {
+    private void testGetFeedbackSessionsPossiblyNeedingClosedEmail() {
         
         ______TS("standard success case");
         
-        List<FeedbackSessionAttributes> fsaList = fsDb.getFeedbackSessionsNeedingClosedEmail();
+        List<FeedbackSessionAttributes> fsaList = fsDb.getFeedbackSessionsPossiblyNeedingClosedEmail();
         
         assertEquals(6, fsaList.size());
         for (FeedbackSessionAttributes fsa : fsaList) {
@@ -219,11 +219,11 @@ public class FeedbackSessionsDbTest extends BaseComponentTestCase {
         
     }
     
-    private void testGetFeedbackSessionsNeedingPublishedEmail() {
+    private void testGetFeedbackSessionsPossiblyNeedingPublishedEmail() {
         
         ______TS("standard success case");
         
-        List<FeedbackSessionAttributes> fsaList = fsDb.getFeedbackSessionsNeedingPublishedEmail();
+        List<FeedbackSessionAttributes> fsaList = fsDb.getFeedbackSessionsPossiblyNeedingPublishedEmail();
         
         assertEquals(8, fsaList.size());
         for (FeedbackSessionAttributes fsa : fsaList) {

--- a/src/test/java/teammates/test/cases/storage/FeedbackSessionsDbTest.java
+++ b/src/test/java/teammates/test/cases/storage/FeedbackSessionsDbTest.java
@@ -100,8 +100,10 @@ public class FeedbackSessionsDbTest extends BaseComponentTestCase {
 
         testGetFeedbackSessions();
         testGetFeedbackSessionsForCourse();
-        testGetFeedbackSessionsWithUnsentOpenEmail();
-        testGetFeedbackSessionsWithUnsentPublishedEmail();
+        testGetFeedbackSessionsNeedingOpenEmail();
+        testGetFeedbackSessionsNeedingClosingEmail();
+        testGetFeedbackSessionsNeedingClosedEmail();
+        testGetFeedbackSessionsNeedingPublishedEmail();
     }
     
     private void testGetFeedbackSessions() {
@@ -176,20 +178,48 @@ public class FeedbackSessionsDbTest extends BaseComponentTestCase {
         assertTrue(fsDb.getFeedbackSessionsForCourse("idOfCourseNoEvals").isEmpty());
     }
     
-    private void testGetFeedbackSessionsWithUnsentOpenEmail() {
+    private void testGetFeedbackSessionsNeedingOpenEmail() {
         
         ______TS("standard success case");
         
         List<FeedbackSessionAttributes> fsaList = fsDb.getFeedbackSessionsNeedingOpenEmail();
         
-        assertEquals(2, fsaList.size());
+        assertEquals(1, fsaList.size());
         for (FeedbackSessionAttributes fsa : fsaList) {
             assertFalse(fsa.isSentOpenEmail());
         }
         
     }
     
-    private void testGetFeedbackSessionsWithUnsentPublishedEmail() {
+    private void testGetFeedbackSessionsNeedingClosingEmail() {
+        
+        ______TS("standard success case");
+        
+        List<FeedbackSessionAttributes> fsaList = fsDb.getFeedbackSessionsNeedingClosingEmail();
+        
+        assertEquals(6, fsaList.size());
+        for (FeedbackSessionAttributes fsa : fsaList) {
+            assertFalse(fsa.isSentClosingEmail());
+            assertTrue(fsa.isClosingEmailEnabled());
+        }
+        
+    }
+    
+    private void testGetFeedbackSessionsNeedingClosedEmail() {
+        
+        ______TS("standard success case");
+        
+        List<FeedbackSessionAttributes> fsaList = fsDb.getFeedbackSessionsNeedingClosedEmail();
+        
+        assertEquals(6, fsaList.size());
+        for (FeedbackSessionAttributes fsa : fsaList) {
+            assertFalse(fsa.isSentClosedEmail());
+            assertTrue(fsa.isClosingEmailEnabled());
+        }
+        
+    }
+    
+    private void testGetFeedbackSessionsNeedingPublishedEmail() {
         
         ______TS("standard success case");
         
@@ -198,6 +228,7 @@ public class FeedbackSessionsDbTest extends BaseComponentTestCase {
         assertEquals(8, fsaList.size());
         for (FeedbackSessionAttributes fsa : fsaList) {
             assertFalse(fsa.isSentPublishedEmail());
+            assertTrue(fsa.isPublishedEmailEnabled());
         }
         
     }

--- a/src/test/java/teammates/test/cases/storage/FeedbackSessionsDbTest.java
+++ b/src/test/java/teammates/test/cases/storage/FeedbackSessionsDbTest.java
@@ -180,7 +180,7 @@ public class FeedbackSessionsDbTest extends BaseComponentTestCase {
         
         ______TS("standard success case");
         
-        List<FeedbackSessionAttributes> fsaList = fsDb.getFeedbackSessionsWithUnsentOpenEmail();
+        List<FeedbackSessionAttributes> fsaList = fsDb.getFeedbackSessionsNeedingOpenEmail();
         
         assertEquals(2, fsaList.size());
         for (FeedbackSessionAttributes fsa : fsaList) {
@@ -193,7 +193,7 @@ public class FeedbackSessionsDbTest extends BaseComponentTestCase {
         
         ______TS("standard success case");
         
-        List<FeedbackSessionAttributes> fsaList = fsDb.getFeedbackSessionsWithUnsentPublishedEmail();
+        List<FeedbackSessionAttributes> fsaList = fsDb.getFeedbackSessionsNeedingPublishedEmail();
         
         assertEquals(8, fsaList.size());
         for (FeedbackSessionAttributes fsa : fsaList) {

